lib/Sema/CSDiag.cpp

```diff
@@ -581,9 +581,10 @@ static bool diagnoseAmbiguity(ConstraintSystem &cs,
 /// This returns the base subexpression that looks immutable (or that can't be
 /// analyzed any further) along with a decl extracted from it if we could.
 ///
-static std::pair<Expr*, ValueDecl*>
-resolveImmutableBase(Expr *expr, ConstraintSystem &CS) {
+static std::pair<Expr *, ValueDecl *>
+resolveImmutableBase(Expr *expr, const Solution &solution) {
```

Comment:
The resolved overloads are, but additional choices are in the solution. In this code (from test/Constraints/lvalue.swift):
```
struct Z {
  subscript(i: Int) -> Double { get {} }
}

var z : Z
z[0] += 0.0 // expected-error{{left side of mutating operator isn't mutable: subscript is get-only}}
```

When `resolveImmutableBase()` is called, there is a disjunction for which subscript member is intended, which hasn't been resolved. Partial CS dump:
```
Inactive Constraints:
  disjunction [[locator@0x10f0a3928 [Subscript@/Users/toon/GitHub/lval.swift:17:2 -> subscript member]]]:($T3) -> $T4 bound to key path application [[locator@0x10f0a3928 [Subscript@/Users/toon/GitHub/lval.swift:17:2 -> subscript member]]]; or ($T3) -> $T4 bound to decl lval.(file).Z.subscript@/Users/toon/GitHub/lval.swift:13:3 : (Int) -> Double at /Users/toon/GitHub/lval.swift:13:3 [[locator@0x10f0a3928 [Subscript@/Users/toon/GitHub/lval.swift:17:2 -> subscript member]]];
  ($T2) arg tuple conv $T3 [[locator@0x10f0a3898 [Subscript@/Users/toon/GitHub/lval.swift:17:2 -> subscript index]]];
Resolved overloads:
  selected overload set choice z: $T1 == @lvalue Z
```

Whereas the solution has chosen the correct overload. Partial solution dump:
```
Overload choices:
  locator@0x10f0a3928 [Subscript@/Users/toon/GitHub/lval.swift:17:2 -> subscript member] with lval.(file).Z.subscript@/Users/toon/GitHub/lval.swift:13:3 as @lvalue Z.subscript: (Int) -> Double

  locator@0x10f0a3760 [DeclRef@/Users/toon/GitHub/lval.swift:17:1] with lval.(file).z@/Users/toon/GitHub/lval.swift:16:5 as z: @lvalue Z

  locator@0x10f0a2e00 [OverloadedDeclRef@/Users/toon/GitHub/lval.swift:17:6] with Swift.(file).Double extension.+= as +=: (inout Double, Double) -> ()
```

The result is that CSDiag.cpp line 600 does not find the get-only member if just the CS is consulted instead of the solution.

Reply:
wait, these are two different things - `findResolvedMemberRef` always uses `getResolvedOverloadSets` which is stored by constraint system. So if there is a single solution (and that's only time when we diagnose fixes at the moment), constraint system should have correct list of resolved overloads...

Reply:
This PR changes the call to `solution.resolveOverloadToDecl()` instead of `CS.findResolvedMemberRef()`. Which uses the solution's `overloadChoices`.

Reply:
What I'm trying to say is that I don't think we need to do that - it should be safe to keep using `findResolvedMemberRef` while trying to diagnose a fix because it should be 1-1 correspondence between `overloadChoices` and `getResolvedOverloadSets` when there is a single solution, if that's not the case there is something else going on...

Reply:
Well, something else is going on then. :) As you can see, both the subscript and the `+=` choices are made in the solution, but aren't in the resolved overloads in the CS.

Reply:
Do you mean that they didn't get printed in `Resolved overloads:` section?
If so, there reason why they didn't is because they haven't been resolved yet (it happens when disjunction choice gets attempted by `resolveOverload`), you should check what the list has after `solve` is done...

Ah, the list is trimmed by `~SolverScope` :,( The reason why it works currently is because a bunch of overloads are resolved upfront by re-typecheck of sub-expressions, this is bad though to rely on that...

I think a good refactoring would be to copy resolved overload set list into solution, that's better than `resolveLocatorToDecl` because `simplifyLocator` is missing a bunch of simplification choices and couldn't be realized upon completely at the moment...